helper/schema/grpc_provider.go

```diff
@@ -1182,6 +1219,8 @@ func (s *GRPCProviderServer) ApplyResourceChange(ctx context.Context, req *tfpro
 
 	newStateVal = copyTimeoutValues(newStateVal, plannedStateVal)
 
+	newStateVal = setWriteOnlyNullValues(newStateVal, schemaBlock)
```

Comment:
I wonder if we should do this in the `ReadResource` RPC as well? It's possible the provider could attempt to refresh the state with a value in a write-only attribute. Terraform core doesn't typically do any validation/assertions then, so there wouldn't be an error.

That seems like being potentially defensive, but we did have a bug about unknown values making it state during refresh recently lol: 
- See outcome and linked issues for more context: https://github.com/hashicorp/terraform-plugin-framework/issues/902

Reply:
Will it be possible for this to affect state? According to the current proposal, write-only attributes will be completely omitted from the state file, so I would be surprised that Terraform Core would write the value even if we do send one across the wire.

Reply:
The current relationship between core and the providers is that the provider is responsible for producing data that satisfies Terraform's constraints. In the case that the provider doesn't, Terraform raises an error diagnostic, rather than silently ignoring the incorrect data.

So if that same principle is applied here, I would expect an error diagnostic if the provider incorrectly sets a write-only attribute. However, we are discussing logic that doesn't exist yet in core, so it's possible they won't take this route. Perhaps we can just note to revisit this when core is closer to implementation